src/llama-context.cpp

```diff
@@ -537,6 +537,17 @@ llm_graph_result_ptr llama_context::build_kv_self_shift(
         const int64_t n_head_kv    = hparams.n_head_kv(il);
         const int64_t n_embd_k_gqa = hparams.n_embd_k_gqa(il);
 
+        float freq_base_l  = cparams.rope_freq_base;
+        float freq_scale_l = cparams.rope_freq_scale;
+
+        // TODO: improve
+        if (model.arch == LLM_ARCH_GEMMA3) {
+            const bool is_sliding = hparams.is_sliding(il);
+
+            freq_base_l  = is_sliding ? 10000.0f : cparams.rope_freq_base;
+            freq_scale_l = is_sliding ? 1.0f     : cparams.rope_freq_scale;
+        }
+
```

Comment:
Not sure how to avoid this special-casing here. It does not look great.

Reply:
I think we can extend the `llama_layer` to hold this info in the near future.

Reply:
For now, I've pushed the following version, which should be a bit cleaner: https://github.com/ggml-org/llama.cpp/pull/12374

Will see if there is a better way to do it with the upcoming model implementation refactoring.